cpp/unreal_plugins/SpEngine/Source/SpEngine/SpEngine.cpp

```diff
@@ -0,0 +1,65 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#include "SpEngine/SpEngine.h"
+
+#include <memory>                  // std::make_unique, std::unique_ptr
+
+#include <Modules/ModuleManager.h> // IMPLEMENT_MODULE
+
+#include "SpCore/Assert.h"
+#include "SpCore/Config.h"
+#include "SpCore/Log.h"
+#include "SpCore/Rpclib.h" // rpc_server
+#include "SpCore/Unreal.h"
+#include "SpEngine/EngineService.h"
+#include "SpEngine/GameWorldService.h"
+#include "SpEngine/LegacyService.h"
+
+void SpEngine::StartupModule()
+{
+    SP_LOG_CURRENT_FUNCTION();
+    SP_ASSERT(FModuleManager::Get().IsModuleLoaded(Unreal::toFName("SpCore")));
+    SP_ASSERT(FModuleManager::Get().IsModuleLoaded(Unreal::toFName("UrdfRobot")));
+    SP_ASSERT(FModuleManager::Get().IsModuleLoaded(Unreal::toFName("Vehicle")));
+
+    if (Config::s_initialized_) {
+        rpc_server_ = std::make_unique<rpc::server>(Config::get<int>("SP_ENGINE.PORT"));
```

Comment:
Can't `rpc_server_` be an object instead of a pointer? I am aware that it isn't default-constructible, but you are hard-coding the port to 30000 in the default case here anyway, so there would be no harm in doing the same thing when you declare the object.

Reply:
In the scenario where we use an object of `rpc::server`, the non-config path works fine. However, when we go through the config path, we try to re-create the server with a new port, and this causes an error and crash if we use the same port as the default one (as there is already a server using that port). To avoid this, I'm using a pointer as this will never encounter that problem.

Reply:
Ah bummer, I bet they're not cleaning up resources properly in their assignment operator. What happens if you do this?

```
rpc_server_.close_sessions();
rpc_server_ = rpc::server(PORT);
```

Reply:
I tried this too, and it threw an error as well.

Reply:
Ah that's evil! What was the exact error?

Reply:
This is the error when I try to do either with running `.close_sessions()` or without.

```bash
LogWindows: Error: === Critical error: ===
LogWindows: Error:
LogWindows: Error: Fatal error!
LogWindows: Error:
LogWindows: Error: Unhandled Exception: 0xe06d7363
LogWindows: Error:
LogWindows: Error: [Callstack] 0x00007ffa97245b0c KERNELBASE.dll!UnknownFunction []
LogWindows: Error: [Callstack] 0x00007ffa6bd76ba7 VCRUNTIME140.dll!UnknownFunction []
LogWindows: Error: [Callstack] 0x00007ff6f3b7e072 SpearSim-Cmd.exe!clmdep_asio::detail::throw_exception<std::system_error>() []
LogWindows: Error: [Callstack] 0x00007ff6f3b82577 SpearSim-Cmd.exe!clmdep_asio::detail::do_throw_error() []
LogWindows: Error: [Callstack] 0x00007ff6f3b80c9f SpearSim-Cmd.exe!clmdep_asio::basic_socket_acceptor<clmdep_asio::ip::tcp,clmdep_asio::socket_acceptor_service<clmdep_asio::ip::tcp> >::bind() []
LogWindows: Error: [Callstack] 0x00007ff6f3b7eb81 SpearSim-Cmd.exe!rpc::server::impl::impl() []
LogWindows: Error: [Callstack] 0x00007ff6f3b7ee28 SpearSim-Cmd.exe!rpc::server::server() []
LogWindows: Error: [Callstack] 0x00007ff6eaa44879 SpearSim-Cmd.exe!SpEngine::StartupModule() [C:\github\spear\cpp\unreal_projects\SpearSim\Plugins\SpEngine\Source\SpEngine\SpEngine.cpp:30]
LogWindows: Error: [Callstack] 0x00007ff6ed44c92b SpearSim-Cmd.exe!FModuleManager::LoadModuleWithFailureReason() []
LogWindows: Error: [Callstack] 0x00007ff6ed7af355 SpearSim-Cmd.exe!FModuleDescriptor::LoadModulesForPhase() []
LogWindows: Error: [Callstack] 0x00007ff6ed7c28d2 SpearSim-Cmd.exe!FPluginManager::TryLoadModulesForPlugin() []
LogWindows: Error: [Callstack] 0x00007ff6ed7aeff4 SpearSim-Cmd.exe!FPluginManager::LoadModulesForEnabledPlugins() []
LogWindows: Error: [Callstack] 0x00007ff6ea8d538a SpearSim-Cmd.exe!FEngineLoop::LoadStartupModules() []
LogWindows: Error: [Callstack] 0x00007ff6ea8d88a5 SpearSim-Cmd.exe!FEngineLoop::PreInitPostStartupScreen() []
LogWindows: Error: [Callstack] 0x00007ff6ea8d0e78 SpearSim-Cmd.exe!GuardedMain() []
LogWindows: Error: [Callstack] 0x00007ff6ea8d11ca SpearSim-Cmd.exe!GuardedMainWrapper() []
LogWindows: Error: [Callstack] 0x00007ff6ea8d40d6 SpearSim-Cmd.exe!LaunchWindowsStartup() []
LogWindows: Error: [Callstack] 0x00007ff6ea8e6e04 SpearSim-Cmd.exe!WinMain() []
LogWindows: Error: [Callstack] 0x00007ff6f45d5d06 SpearSim-Cmd.exe!__scrt_common_main_seh() [D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288]
LogWindows: Error: [Callstack] 0x00007ffa985e257d KERNEL32.DLL!UnknownFunction []
LogWindows: Error: [Callstack] 0x00007ffa9990aa58 ntdll.dll!UnknownFunction []
LogWindows: Error:
LogExit: Executing StaticShutdownAfterError
LogWindows: FPlatformMisc::RequestExit(1)
LogWindows: FPlatformMisc::RequestExitWithStatus(1, 3)
LogCore: Engine exit requested (reason: Win RequestExit)
Traceback (most recent call last):
  File ".\run_executable.py", line 128, in <module>
    subprocess.run(cmd, check=True)
  File "C:\Users\Rachith\miniconda3\envs\spear\lib\subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\github\\spear\\cpp\\unreal_projects\\SpearSim\\Standalone-Development\\Windows\\SpearSim\\Binaries\\Win64\\SpearSim-Cmd.exe', '-windowed', '-resx=1024', '-resy=1024', '-graphicsadapter=0', '-nosound', '-fileopenlog', '-stdout', '-fullstdoutlogoutput', '-nologtimes', '-config_file=C:\\github\\spear\\tools\\tmp\\config.yaml']' returned non-zero exit status 3.

```

Reply:
And, this is the C++ compiler version I'm seeing
```
> C:\Program Files\Microsoft Visual Studio\2022\Community>cl
Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31948 for x64
```

Reply:
Bummer.